cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -150,6 +157,85 @@ Result<std::unique_ptr<KernelState>> CountInit(KernelContext*,
   return std::make_unique<CountImpl>(static_cast<const CountOptions&>(*args.options));
 }
 
+// ----------------------------------------------------------------------
+// Distinct implementations
+
+struct DistinctImpl : public ScalarAggregator {
+  Status Consume(KernelContext* ctx, const ExecSpan& batch) override {
+    if (batch[0].is_array()) {
+      const ArraySpan& input = batch[0].array;
+      this->arrays.push_back(input.ToArray());
+    } else {
+      const Scalar& input = *batch[0].scalar;
+      std::shared_ptr<arrow::Array> scalar_array;
+      ARROW_ASSIGN_OR_RAISE(scalar_array,
+                            arrow::MakeArrayFromScalar(input, 1, ctx->memory_pool()));
+      this->arrays.push_back(scalar_array);
+    }
+    return Status::OK();
+  }
```

Comment:
This approach is "ok" but I think we can do better.  It will accumulate the entire array in memory.  We could instead, compute distinct items here and only add those that aren't already in the set.

We could use a memo table. In consume we could update the memo table with the new values.  In the end we just dump out the memo table.

Reply:
@westonpace I used an alternative method, could you please take a look?

Reply:
On second thoughts, I think this method would need tweaks for Scalar values separately. Let me work on this a bit further.